File: src/main/java/me/ghostbear/kumaslash/data/tachiyomi/Extension.java
package me.ghostbear.kumaslash.data.tachiyomi;

import java.util.Set;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.PersistenceCreator;
import org.springframework.data.annotation.Transient;
import org.springframework.data.domain.Persistable;
import org.springframework.data.relational.core.mapping.Column;
import org.springframework.data.relational.core.mapping.Table;

@Table("extension")
public record Extension(
		@Id
		@Column("package_name")
		String packageName,
		@Column("name")
		String name,
		@Column("file_name")
		String fileName,
		@Column("language")
		String language,
		@Column("code")
		long code,
		@Column("version")
		String version,
		@Column("is_nsfw")
		boolean isNsfw,
		@Column("has_readme")
		boolean hasReadme,
		@Column("has_changelog")
		boolean hasChangelog,
		@Transient
		Set<Source> sources,
		@Transient
		boolean isNew
) implements Persistable<String> {

	@PersistenceCreator
	public Extension(String packageName, String name, String fileName, String language, long code, String version, boolean isNsfw, boolean hasReadme, boolean hasChangelog) {
		this(packageName, name, fileName, language, code, version, isNsfw, hasReadme, hasChangelog, null, false);
	}

	public Extension(String packageName, String name, String fileName, String language, long code, String version, boolean isNsfw, boolean hasReadme, boolean hasChangelog, Set<Source> sources) {
		this(packageName, name, fileName, language, code, version, isNsfw, hasReadme, hasChangelog, sources, false);
	}

	@Override
	public String getId() {
		return packageName;
	}

	public Extension withIsNew(boolean isNew) {
		return new Extension(packageName, name, fileName, language, code, version, isNsfw, hasReadme, hasChangelog, sources, isNew);
	}
}

File: src/main/kotlin/me/ghostbear/kumaslash/data/tachiyomi/Extension.kt
package me.ghostbear.kumaslash.data.tachiyomi

import dev.kord.rest.builder.component.ActionRowBuilder
import dev.kord.rest.builder.message.modify.InteractionResponseModifyBuilder
import kotlinx.serialization.Serializable

@Serializable
data class Extension(
    val name: String,
    val pkg: String,
    val apk: String,
    val lang: String,
    val code: Long,
    val version: String,
    val nsfw: Long,
    val hasReadme: Long,
    val hasChangelog: Long,
    val sources: List<Source>
)

fun Extension.toMessage(): InteractionResponseModifyBuilder.() -> Unit = {
    content = "<:nadeshikosuccess:843281745088217158> Source was found in the **${name.replace("Tachiyomi: ", "")}** extension."
    components = (components ?: mutableListOf())
    components!!.add(
        ActionRowBuilder().apply {
            linkButton("https://raw.githubusercontent.com/tachiyomiorg/tachiyomi-extensions/repo/apk/$apk") {
                label = "Download"
            }
        }
    )
}
